### DISCUSSION TITLE
Project scope

### DISCUSSION THREAD

**Discussion by ReblochonMasque** [rvodden/pysketcher#318] [General]
Hello Richard,

Tremendous work on reorganizing pysketcher! 

I've been attracted on and off by this project for some years; I think it would be useful at this time to have a clear scope formalized for pysketcher. Maybe something a little bit more precise and detailed than the "elevator pitch" found in the readme, and the about sections. Something that would help a reader/contributor see a little bit into the future.

**about:**
> Python-based drawing tool for making sketches of mechanics problems. 

**readme:**
> PySketcher is a modern Python library designed to make creating geometric, mathematical and physical diagrams very easy.

**Comment by rvodden** [comment]
There is a clear immediate scope in the issues which are tagged with the 0.1.0 release. At this point, we'll be feature-complete with HPLs original work. Once we're there I'd like to add a geometric solving capability, which is clear from the other discussion. Is there anything else you were looking for in terms of direction?

**Comment by ReblochonMasque** [comment]
Thank you!
What sort of geometric problems have you got in mind?
Same for mechanics and physics: what do you want pysketcher to be able to achieve?
It is not CAD, nobody will ever use it for industrial drawings. It is not a numerical solver, no-one will use it for serious simulations. It is not interactive, so it is not an intuitive tool you can play with to explore a domain. 
These are important questions - not only for a potential contributor like myself, but for you to clarify your thoughts. 
Maybe it is a support for articles on your blog, or maybe it is something else... IDK, it is your responsibility to clarify!
A careful scoping is needed to enable pysketcher to find its way... Maybe you already know, but I am not a mind reader.

As far as I see it, there is a base that can be built upon, but before that, it needs serious consolidating:
The project conflates `Point` and `Vector`; I guess it is okay to place simple things on a screen, but we'll need vectors to solve problems. Coordinate Systems are implicit, which is also not great to solve physics problems. `Line` is defined using a high school equation that cannot handle vertical; `Gravity` inherits from `Force` (it is an acceleration); `Shape` would probably benefit from a recursive structure that can enable assemblies; this is not an easy first contribution changing this will touch every class in the project. There are two classes that handle `COLOR`; each uses its own code for the same colors (possibly easy to fix). Unlike what the readme blurb pretends, the entire project is tightly coupled to matplotlib... and that's okay, we just need to acknowledge it, and either move towards more matplotlib, or away towards a variety of backends.
There are more issues to address, but this sampling is sufficient to illustrate why I think we will stay in alpha for a bit more time than anticipated.

Your turn! :)

**Comment by rvodden** [comment]
The initial direction is the 0.1.0 release which will see feature parity with the original library. I agree that there are some existential questions that need to be answered, but I think they can be considered as we work towards 0.1.0. Pysketcher was originally written to help HPL produce the diagrams for one of his books. That's where I see us providing the most value. Illustrative diagrams which support educational/informational texts. So for example, we might implement a geometric solver so that it's much easier for a mathematics professor to provide an illustration for her examples, but (as you quite correctly say) we're not going to produce an arbitrary precision numerical simulator. 

I certainly appreciate the challenges you've presented, and they certainly need solving. I'll work through them, and anything which prevents us presenting a stable API at 0.1.0 we'll add to the scope of that release. Note I say a stable API, not one that will not change. That means we can add stuff (e.g. Vector) but not take stuff away. To take stuff away we'd need to deprecate it and then remove it in a subsequent release - perfectly possible, but very annoying to have to do until that stability has been achieved.

1) **Point and Vector are conflated** - I totally agree, I have considered adding a Vector class. I've held back as that's not in HPL's original code. In my head we've been working towards re-implementing all of HPLs functionality (or explicitly deciding not to) and only then adding new stuff. I think this deserves an issue, I don't think we need to fix it for 0.1.0, but that doesn't mean we can't include it if its ready. We'll need to go through all the examples and ensure that the right class is used as part of that PR.

1) **Co-ordinate systems are implicit**. Yeah agree with this too. Would be great if you could illustrate the kinds of problems which require (or would be helped by having) an explicit coordinate system in order to illustrate them. That will help guide how we might resolve this. Again spinning up an issue for this would be helpful.

1) **Line is defined using a high school equation that cannot handle vertical**. If vertical doesn't work please raise a bug :-) The equation for a line shouldn't be in the Line object at all. This is an HPL legacy as his code had every object be a specialization of a curve, which effectively means that every object knows how to draw itself. This makes sense if all we're going to do is implement a matplotlib backend, as that only has Curve as a primitive. It makes no sense at all in the context of, say, an SVG backend where there are many more primitives. #17 is open to address this, but needs some thought. There are other issues like this too, for example a Rectangle is a composition of 4 lines, which it only needs to be if that is how the backend chooses to render it, but we don't want to repeat the "convert my rectangle to 4 lines" code in every backend which chooses to render it that way. I need to chew on the class structure a bit before diving in. I've just noticed that #17 has "good first issue" tagged on it, which is obviously not the case, so I've removed it.

1) **Gravity inherits from Force** - this does not assert that Gravity and Force are the same things, it asserts that they are represented by a specialisation of the same symbol.

1) **Shape would probably benefit from a recursive structure that can enable assemblies; this is not an easy first contribution changing this will touch every class in the project.** - What problem would this solve? I think I understand, but I don't want to assume that I've read your mind in the same way you don't want to assume that you've read mine :-)

1) **There are two classes that handle COLOR; each uses its own code for the same colors (possibly easy to fix).** There almost certainly shouldn't be. Can you raise an issue and expand on this a bit? We should either fix it, or not use the same term twice.

1) **Unlike what the readme blurb pretends, the entire project is tightly coupled to matplotlib... ** I didn't say that the project doesn't need matplotlib. What I said was that it is no longer tightly coupled to the shape obect. The original code had a global matplotlib object and matplotlib code in each shape class. The decoupled statement simply means that a new backend is possible, not that we've actually done it yet. I do think implementing one would be a good idea though - otherwise we don't really know if its possible or not. I suggest that #17 is fixed first, however.

**There are more issues to address, but this sampling is sufficient to illustrate why I think we will stay in alpha for a bit more time than anticipated.** I've not set a timeline on 0.1.0; and I've not said that we'll leave alpha at 0.1.0 either. All I've said is that we won't release 0.1.0 until the API is stable. That might mean we're at version 0.0.600 in 7 years time (although I hope not!!).

This is a really great discussion - thanks for prompting it.

P.S. I'm not overly interested as all that matters for this conversation is how it's drawn, but my understanding is that gravity is very much a force. For example, if I have a brick sitting on a tabletop and the tabletop is able to support the brick's weight then there is no acceleration; the system is in equilibrium. This is because the gravitational _force_ is entirely counteracted by the reaction force of the table. There is no acceleration, but there is gravity so they cannot be the same thing. As I say, that's just my understanding - I am no physicist. I'd be fascinated to understand your assertion that gravity is an acceleration, particularly in my apparent counterexample - as I might end up understanding things a little better.

**Comment by ReblochonMasque** [comment]
Great, thanks, this is wonderful - I will need a bit of time to chew on it... and your constraints articles that I am trying to understand. 

Your PS:
Gravity is expressed in `[L]/[T]^2` - this is a dimensional equation that means units of Length, divided by units of time, squared; this could be `m/s^2`, `ft/hour^2`, it does not matter: if its dimension is acceleration, it is an acceleration.
If an object does not accelerate, its acceleration is zero, that means that no net force and no net moments are acting on it; it is in what physicists and engineers call a state of **Static Equilibrium**. It is one of the major applications of Newton's first and third laws:

- A particle at rest, or moving in a straight line at constant velocity remains in this state provided it is not subject to an unbalanced force.
- The mutual forces of action and reaction between 2 particles are equal, opposite, and collinear. (your example of the brick on a tabletop: the weight, which is mass times gravity (a Force), is balanced by the reaction force of the table, therefore the brick does not move)

Dimensional equation: `[Force] = [Mass] * [acceleration] =  [Mass]*[Length]/[Time]^2` that is one Newton (a unit of force) is, expressed in the International System of Units a `kg m / s^2`

In my opinion, how it is drawn is a consequence to how precise, and how carefully we can define these concepts and represent them. We cannot pretend solving physics problems without being quite formal in our definitions, and notations. In other words: the calculator/solver must probably be entirely decoupled from the way things are drawn, take its input from the drawing, calculate things, then give its output to be rendered.

**Reply by rvodden** [reply]
Ooo thanks :-) That makes total sense as my brick on its table also has mass!

**Reply by ReblochonMasque** [reply]
Yes, `weight = mass * gravity` - weight is a force.
